File: Top_Interview_Questions/8_StringToInteger.cpp
class Solution {
public:
    int myAtoi(string str) {
        if(str.empty()) return 0;
        
        int cur = 0;
        while(cur < str.size() && str[cur] == ' ')
            ++cur;
        int sign = 1;// 1 +, -1 -
        if(str[cur] == '-') {
            sign = -1;
            ++cur;
        } else if(str[cur] == '+')
            ++cur;
        
        int ans = 0;
        while(cur < str.size() && isNum(str[cur])) {
            //同时需要判断末位是否溢出
			if(ans > INT_MAX / 10 || (ans == INT_MAX / 10 && str[cur] - '0' > 7)) return sign == -1 ? INT_MIN : INT_MAX;
            
            ans = ans * 10 + (str[cur] - '0');//这里需要括号，否则*10 + ASCII可能会溢出
            ++cur;
        }
        return ans * sign;
    }
private:
    bool isNum(const char c) {
        return c >= '0' && c <= '9';
    }
};

File: Question_Java/8_String_to_Integer.java
class Solution {
    public int myAtoi(String s) {
        if (s.isEmpty()) return 0;
        char[] chs = s.toCharArray();
        int sign = 1;
        int cur = 0;
        // 跳空格
        for (; cur < chs.length; ++cur) {
            if (chs[cur] != ' ') break;
        }
        if (cur < chs.length && (chs[cur] == '-' || chs[cur] == '+')) {
            sign = chs[cur] == '-' ? -1 : 1;
            ++cur;
        }

        int ans = 0;
        while (cur < chs.length) {
            if (chs[cur] > '9' || chs[cur] < '0') return ans * sign;
            int num = chs[cur] - '0';
            if (ans > Integer.MAX_VALUE / 10) return sign == -1 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            if (ans == Integer.MAX_VALUE / 10) {
                if (sign == 1 && num >= 7) return Integer.MAX_VALUE;
                if (sign == -1 && num >= 8) return Integer.MIN_VALUE; 
            }
            ans = ans * 10 + num;
            ++cur;
        }
        return ans * sign;
    }
}